linked_list/linked_list.py

```diff
@@ -2,116 +2,234 @@
+        self.head = None  # keep the head private. Not accessible outside this class
+        self.tail = None
 
     # returns the value in the first node
     # returns None if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def get_first(self):
-        pass
+        if self.head is None:
+            return None
 
+        return self.head.value
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
+
     def add_first(self, value):
-        pass
+        if self.head is None:
+            self.head = self.tail = Node(value)
+        else:
+            new_node = Node(value, self.head)
+            self.head.previous = new_node
+            self.head = new_node
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        if self.head is None:
+            return False
+
+        current = self.head
+        while current is not None:
+            if current.value == value:
+                return True
+            else:
+                current.previous = current
+                current = current.next
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        list_length = 0
+        current = self.head
+
+        while current is not None:
+            list_length += 1
+            current = current.next
+
+        return list_length
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        current = self.head
+        current_index = 0
+        if index < 0 or current is None:
+            return None
+
+        while current_index <= index:
+            if current_index != index:
+                current = current.next
+                current_index +=1
+            elif current_index == index:
+                return current.value
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
+
     def get_last(self):

```

Comment:
👍  Since you have a doubly linked list with a `tail` attribute you can do this in O(1) time complexity.